Initial implementation of lazy JLLs for LinearAlgebra

staticfloat opened the pull request:
This alters CompilerSupportLibraries_jll, OpenBLAS_jll and
libblastrampoline_jll to use `LazyLibrary` objects and thereby be loaded
only upon first `dlopen()` or `ccall()` to the individual library
objects.  Note that this is one of the more complicated cases, as
`libblastrampoline` must have OpenBLAS_jll added as a dynamic dependency
(as it does not actually have it listed in its shared object headers)
and also has some on-load callbacks that must be invoked.

Long-term, I would like to replace the bespoke JLLs here with actual JLL
source code, and vendor a version of `LazyJLLWrappers` in-tree to do the
actual code generation even for Base.  That is left as future work.

[0] https://github.com/JuliaLang/LinearAlgebra.jl/pull/1235

staticfloat committed:
Initial implementation of lazy JLLs for LinearAlgebra

This alters CompilerSupportLibraries_jll, OpenBLAS_jll and
libblastrampoline_jll to use `LazyLibrary` objects and thereby be loaded
only upon first `dlopen()` or `ccall()` to the individual library
objects.  Note that this is one of the more complicated cases, as
`libblastrampoline` must have OpenBLAS_jll added as a dynamic dependency
(as it does not actually have it listed in its shared object headers)
and also has some on-load callbacks that must be invoked.

Long-term, I would like to replace the bespoke JLLs here with actual JLL
source code, and vendor a version of `LazyJLLWrappers` in-tree to do the
actual code generation even for Base.  That is left as future work.

This must be paired with the appropriate `LinearAlgebra.jl` changes [0].

[0] https://github.com/JuliaLang/LinearAlgebra.jl/pull/1235

staticfloat commented:
This is working locally now, I can confirm that `Libdl.dllist()` shows that `libgfortran`, `libopenblas64_` and `libblastrampoline` are all not loaded until the first Linear Algebra call!

Note that this is attempting to conform to [the `LazyJLLWrappers` API](https://github.com/JuliaPackaging/BinaryBuilder2.jl/tree/main/LazyJLLWrappers.jl), which includes things like `eager_mode()`, which is [called by `JLLWrappers.jl`](https://github.com/JuliaPackaging/JLLWrappers.jl/blob/ecaba50a4462209714f0979667c64c1bf28ee892/src/wrapper_generators.jl#L40-L42).  This is to provide backwards-compatibility with JLLs that expect all dependencies from all dependent JLLs to have already `dlopen()`'ed their libraries.

What this means is that for any realistic workload where there are JLLs that depend on LBT/OpenBLAS, everything is going to get eagerly loaded upon loading of that JLL anyway, until those JLLs are built to use `LazyJLLWrappers` instead of `JLLWrappers`, which functionally means waiting for [BB2](https://github.com/JuliaPackaging/BinaryBuilder2.jl) to be ready, as it contains the new analysis passes that track things like library interdependencies, etc...  However, for workloads that _never_ use linear algebra workloads, this will be a nice savings, as the libraries will now never get loaded!

topolarity commented:
The [`--trim` failure](https://buildkite.com/julialang/julia-master/builds/45690#01958705-b3f7-4b1a-9546-1773ecf0539f/800-848) should be fixed up by https://github.com/JuliaLang/julia/pull/57734

staticfloat commented:
I've converted this to draft because I realized that while "real" JLLs that use `JLLWrappers` will correctly call `eager_mode()` on these lazy stdlib JLLs, other stdlib JLLs that are not lazy yet but that depend on any of these new lazy stdlib JLLs won't, I need to add that in here.

topolarity commented:
It's also worth noting that once we land this PR, `--trim` will not support loading BLAS until we fix https://github.com/JuliaLang/julia/issues/57707

staticfloat committed:
Change `BundledLazyLibraryPath()` to be relative to `private_shlibdir`

This makes more sense, reduces work, and avoids that perennial thorn in
my side, that in-tree builds store libraries in `${prefix}/lib`, whereas
installed builds store libraries in `${prefix}/lib/julia`.

Co-authored-by: Cody Tapscott <84105208+topolarity@users.noreply.github.com>

topolarity commented on base/libdl.jl:
```diff
@@ -339,9 +339,15 @@ Base.cconvert(::Type{Cstring}, llp::LazyLibraryPath) = Base.cconvert(Cstring, st
-Base.string(::SysBindirGetter) = dirname(Sys.BINDIR)
+# Helper to get `$(private_shlibdir)` at runtime
+struct PrivateShlibdirGetter; end
+const private_shlibdir::Ref{String} = Ref{String}()
+function Base.string(::PrivateShlibdirGetter)
+    if !isassigned(private_shlibdir)
+        private_shlibdir[] = dirname(only(filter(p -> contains(p, "libjulia-internal"), dllist())))
+    end
+    return private_shlibdir[]
+end
```
```suggestion
const private_shlibdir = Base.OncePerProcess{String}() do
    dirname(only(filter(p -> contains(p, "libjulia-internal"), dllist())))
end
Base.string(::PrivateShlibdirGetter) = private_shlibdir()
```

`OncePerProcess` has the benefit of being thread-safe and resetting after pre-compilation

That protects against the case that the libdir could "stick" from the build machine if you look it up during pre-compilation IIUC

topolarity commented on stdlib/CompilerSupportLibraries_jll/src/CompilerSupportLibraries_jll.jl:
```diff
@@ -5,80 +5,108 @@
+    dlopen(libgfortran)
+    dlopen(libstdcxx)
+end
+is_available() = true
```
Out of ignorance, what is this check actually used for? Is it OK that this "lies" if the libraries aren't around to be lazily loaded?

staticfloat replied:
This is used to determine if a JLL has a build at all, not that all possible libraries are present. For stdlib JLLs they are always true by construction, except for cases like libunwind perhaps?

topolarity commented:
I think you might need a rebase for the `trimming` test to pass, and the following patch:
```diff
diff --git i/base/libdl.jl w/base/libdl.jl
index 77f608e77e..d94a2fd0e2 100644
--- i/base/libdl.jl
+++ w/base/libdl.jl
@@ -333,7 +333,7 @@ struct LazyLibraryPath
     pieces::Tuple{Vararg{Any}}
     LazyLibraryPath(pieces...) = new(pieces)
 end
-Base.string(llp::LazyLibraryPath) = joinpath(String[string(p) for p in llp.pieces])
+@inline Base.string(llp::LazyLibraryPath) = joinpath(String[string(p) for p in llp.pieces])
 Base.cconvert(::Type{Cstring}, llp::LazyLibraryPath) = Base.cconvert(Cstring, string(llp))
 # Define `print` so that we can wrap this in a `LazyString`
 Base.print(io::IO, llp::LazyLibraryPath) = print(io, string(llp))
```

The robust solution to this will require a callable of some kind, but for now massaging is required to make sure that the components of the `path` are visible to the optimizer

staticfloat committed:
Teach `detect_{libgfortran,libstdcxx}_version` to deal with lazy CSL

We can no longer depend on the fact that these libraries will be eagerly
loaded (huzzah!) so we must teach these helper functions to either
inspect the lazy CSL JLL for the necessary information, or just directly
load the library.

staticfloat committed:
test/trimming: Use `SHLIB_EXT` in Makefile

staticfloat committed:
Bump LinearAlgebra version to latest

This includes the changes needed for lazy LinearAlgebra JLLs.

staticfloat commented:
Tests passed, so I merged the `LinearAlgebra` PR, adjusted the final commit here to bump the LinearAlgebra stdlib, and once this goes green, we're good to merge!

topolarity commented:
@nanosoldier `runtests(vs = "@04922e7e205237ab01cb009b2684bcbb97b053cf")`

giordano commented:
#57808

topolarity commented:
Fix in https://github.com/JuliaLang/julia/pull/57811/

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/57719#issuecomment-2730899827) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f38a3b9_vs_04922e7/report.html) is available.

<details><summary>Report summary</summary>

## ❗ Packages that crashed

**1 packages crashed only on the current version.**

 - An internal error was encountered: 1 packages

4 packages crashed on the previous version too.

## ✖ Packages that failed

**31 packages failed only on the current version.**

 - Package fails to precompile: 5 packages
 - Illegal method overwrites during precompilation: 2 packages
 - Package has test failures: 5 packages
 - Package tests unexpectedly errored: 2 packages
 - Package requires a missing binary dependency: 4 packages
 - There were unidentified errors: 1 packages
 - Test duration exceeded the time limit: 12 packages

1518 packages failed on the previous version too.

## ✔ Packages that passed tests

**21 packages passed tests only on the current version.**

 - Other: 21 packages

5084 packages passed tests on the previous version too.

## ~ Packages that at least loaded

**7 packages successfully loaded only on the current version.**

 - Other: 7 packages

2817 packages successfully loaded on the previous version too.

## ➖ Packages that were skipped altogether

**1 packages were skipped only on the current version.**

 - Package could not be installed: 1 packages

909 packages were skipped on the previous version too.



</details>

topolarity commented:
Some of the PkgEval failures (thankfully, not many!) look relevant, such as [FastPolynomialRoots.jl](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f38a3b9_vs_04922e7/FastPolynomialRoots.primary.log) and [StateSpaceLearning.jl](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/f38a3b9_vs_04922e7/StateSpaceLearning.primary.log):
```
Failed to precompile StateSpaceLearning [971c4b7c-2c4e-4bac-8525-e842df3cde7b] to "/home/pkgeval/.julia/compiled/v1.13/StateSpaceLearning/jl_cmLoFo" (ProcessExited(1)).
ERROR: LoadError: InitError: could not load library "/home/pkgeval/.julia/artifacts/e80442d59af73a1877a108eb3fab2cd12d25dc56/lib/libglmnet.so"
libgfortran.so.5: cannot open shared object file: No such file or directory
```

topolarity commented:
Looks like the common theme is that the JLL's involved are so old (≥5 years) that they [predate JLLWrappers](https://github.com/JuliaBinaryWrappers/glmnet_jll.jl/blob/dafec0cd7d988a584f1c33ec9b73c8d26054d6b8/src/wrappers/x86_64-linux-gnu-libgfortran5.jl#L37)

gitboy16 commented:
Just a note that between this PR and the previous one 57752, the size of Julia windows installation folder jumped from 1.1G to 1.5G.

KristofferC commented:
I removed the backport here because this is very much like a feature and feature freeze was many months ago.

topolarity commented:
> Just a note that between this PR and the previous one 57752, the size of Julia windows installation folder jumped from 1.1G to 1.5G.

Any idea which files / folders experienced the difference?
